### PR TITLE
Add bfloat16 floating-point format support based on AMP 

### PR DESCRIPTION
**Pull request by rongzha1**
## Description ##
Bfloat16 is wildly used in Deep Learning especially on training to get a better performance.

This PR is to add bf16 support based on mxnet AMP(Automatic Mixed Precision) module .

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [done ] Changes are complete (i.e. I finished coding on this PR)
- [done ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at https://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ done] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

This PR has passed unitest and preci test in local machine.
Unit tests are added for this PR.

@ZhennanQin @ElaineBao @xinyu-intel @TaoLv @PatricZhao

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -6,7 +6,7 @@
 	url = https://github.com/dmlc/ps-lite
 [submodule "3rdparty/dlpack"]
 	path = 3rdparty/dlpack
-	url = https://github.com/dmlc/dlpack
+	url = https://github.com/ElaineBao/dlpack.git
```
Comment (ptrendx) [inline .gitmodules]: Will need to be changed once the changes are merged to upstream dlpack. Leaving this comment as a reminder ;-).

File: 3rdparty/mshadow/mshadow/base.h [hunk 2]
```diff
@@ -312,6 +338,11 @@ enum TypeFlag {
   kInt8  = 5,
   kInt64 = 6,
   kBool = 7,
+  kInt16 = 8,
+  kUint16 = 9,
+  kUint32 = 10,
+  kUint64 = 11,
```
Comment (ptrendx) [inline 3rdparty/mshadow/mshadow/base.h]: Why adding those additional types here? No operator supports them anyway, right?

File: 3rdparty/mshadow/mshadow/bfloat.h [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    return RTYPE(static_cast<float>(a) OP float(b));  /* NOLINT(*) */     \
+  }                                                                       \
+  MSHADOW_BF16_OPERATOR_TYPE(float, float, OP)                            \
+  MSHADOW_BF16_OPERATOR_TYPE(double, double, OP)                          \
+  MSHADOW_BF16_OPERATOR_TYPE(float, int8_t, OP)                           \
+  MSHADOW_BF16_OPERATOR_TYPE(float, uint8_t, OP)                          \
+  MSHADOW_BF16_OPERATOR_TYPE(float, int32_t, OP)                          \
+  MSHADOW_BF16_OPERATOR_TYPE(float, uint32_t, OP)                         \
+  MSHADOW_BF16_OPERATOR_TYPE(float, int64_t, OP)                          \
+  MSHADOW_BF16_OPERATOR_TYPE(float, uint64_t, OP)
```
Comment (ptrendx) [inline 3rdparty/mshadow/mshadow/bfloat.h]: Returning float or double, while understandable, is a different behavior to the one currently done for half_t type. Could we discuss this and make them consistent?

File: example/quantization/imagenet_inference.py [hunk 4]
```diff
@@ -99,7 +100,34 @@ def score(sym, arg_params, aux_params, data, devs, label_name, max_num_examples,
...
+        elif model_name.find('vgg') != -1:
+            excluded_sym_names += ['vgg0_conv0_fwd']
+        elif model_name.find('squeezenet1') != -1:
+            excluded_sym_names += ['squeezenet0_conv0_fwd']
+        elif model_name.find('mobilenet') != -1 and model_name.find('v2') == -1:
+            excluded_sym_names += ['mobilenet0_conv0_fwd']
+        elif model_name.find('mobilenet') != -1 and model_name.find('v2') != -1:
+            excluded_sym_names += ['mobilenetv20_conv0_fwd']
+        elif model_name.find('inceptionv3') != -1:
+            excluded_sym_names += ['inception30_conv0_fwd']
```
Comment (ptrendx) [inline example/quantization/imagenet_inference.py]: Why? Is there an accuracy issue without those exclusions?

File: python/mxnet/contrib/amp/amp.py [hunk 5]
```diff
@@ -43,14 +44,17 @@
 from ... import optimizer as opt
 from .loss_scaler import LossScaler
 
+bfloat16 = np.dtype([('bfloat16', np.uint16)])
```
Comment (ptrendx) [inline python/mxnet/contrib/amp/amp.py]: Can we have this dtype accessible (as `mx.bfloat16` or something similar)?

File: python/mxnet/gluon/parameter.py [hunk 6]
```diff
@@ -289,11 +289,18 @@ def _load_init(self, data, ctx, cast_dtype=False, dtype_source='current'):
...
+                    "Failed loading Parameter '%s' from saved params: " \
+                    "dtype incompatible expected %s vs saved %s. " \
+                    "Set cast_dtype=True to cast the dtype of saved params."%(
+                        self.name, str(self.dtype), str(data.dtype))
+                else:
+                    assert np.dtype(self.dtype).type == data.dtype, \
+                    "Failed loading Parameter '%s' from saved params: " \
+                    "dtype incompatible expected %s vs saved %s. " \
+                    "Set cast_dtype=True to cast the dtype of saved params."%(
+                        self.name, str(self.dtype), str(data.dtype))
```
Comment (ptrendx) [inline python/mxnet/gluon/parameter.py]: Aren't those 2 codepaths the same?

File: 3rdparty/mshadow/mshadow/base.h (hunk 2 above)
Comment (ZhennanQin) [reply 3rdparty/mshadow/mshadow/base.h]: This is to align the definition with DLPack. Otherwise we have to preserve those numbers. Even through we don't use them currently, it's no harm to add them.

File: 3rdparty/mshadow/mshadow/bfloat.h (hunk 3 above)
Comment (ZhennanQin) [reply 3rdparty/mshadow/mshadow/bfloat.h]: Sure. Any suggestion here?

File: example/quantization/imagenet_inference.py (hunk 4 above)
Comment (ZhennanQin) [reply example/quantization/imagenet_inference.py]: Not for accuracy, but for performance purpose. This could be removed once more bfloat16 hardware added.

File: python/mxnet/contrib/amp/amp.py (hunk 5 above)
Comment (ZhennanQin) [reply python/mxnet/contrib/amp/amp.py]: This is a good topic, and I want to have a discussion for this. 
Currently, MXNet doesn't have its own type system. It's simply using Numpy.dtype. Numpy doesn't natively support bfloat16, so we define bfloat16 as a numpy customized type.
Pros: compatible with current design, isinstance(bfloat16, np.dtype) could return True.
cons: bfloat16.name doesn't work, have to use bfloat16.names[0] instead.
Another solution is, creating MXNet's own data type system, just like pytorch and tf. This is a big API change, so we wish this can be done when upgrading to MXNet 2.0.

Currently, we prefer this approach to enable bfloat16 in MXNet 1.x, and refactor it in MXNet 2.0.

File: include/mxnet/ndarray.h [hunk 7]
```diff
@@ -770,6 +770,12 @@ class NDArray {
    */
   NDArray Reorder2Default() const;
 
+    /*
+   * This creates a new NDArray using f32 with the reordered data.
+   * It doesn't affect the data of the original NDArray.
+   */
+  NDArray Reorder2DefaultFp32() const;
```
Comment (eric-haibin-lin) [inline include/mxnet/ndarray.h]: Adding dtype specific interface looks very ad-hoc

File: python/mxnet/ndarray/ndarray.py [hunk 8]
```diff
@@ -83,6 +84,7 @@
     5: np.int8,
     6: np.int64,
     7: np.bool_,
+    12: np.dtype([('bfloat16', np.uint16)]),
```
Comment (eric-haibin-lin) [inline python/mxnet/ndarray/ndarray.py]: why 12?

File: src/engine/naive_engine.cc [hunk 9]
```diff
@@ -55,7 +55,7 @@ class NaiveEngine final : public Engine {
     std::vector<VarHandle> const_vars;
     std::vector<VarHandle> mutable_vars;
     FnProperty prop;
-    const char* opr_name;
+    std::string opr_name;
```
Comment (eric-haibin-lin) [inline src/engine/naive_engine.cc]: Was this an issue blocking this feature? just curious

File: python/mxnet/ndarray/ndarray.py (hunk 8 above)
Comment (ElaineBao) [reply python/mxnet/ndarray/ndarray.py]: This is to align the `TypeFlag` defined in mshadow

File: include/mxnet/ndarray.h (hunk 7 above)
Comment (rongzha1) [reply include/mxnet/ndarray.h]: OK, will change it to Reorder2DefaultFloatFormat()

File: .gitmodules (hunk 1 above)
Comment (pengzhao-intel) [reply .gitmodules]: Definitely :) We're working on PR the related code in dlpack https://github.com/dmlc/dlpack/issues/45

File: src/engine/naive_engine.cc (hunk 9 above)
Comment (ZhennanQin) [reply src/engine/naive_engine.cc]: This is a bugfix for naive engine. If my remember is correct, MXNet already had this fixed with different approach. So we can drop this change from this PR.

Comment (eric-haibin-lin) [reply src/engine/naive_engine.cc]: I don't think it's merged yet. I'm ok with the change here.

**Comment by ElaineBao** [comment]
> For bfloat16 training, which loss scalar is recommended? Do we also need to perform NaN checks?

Bfloat16 has the same dynamic range as float32, since they have the same exponent bits. So it can represent gradients directly, it doesn't require loss scaling like fp16.

File: example/quantization/imagenet_inference.py (hunk 4 above)
Comment (pengzhao-intel) [reply example/quantization/imagenet_inference.py]: Please add a comment for this temp performance solution and will convert all conv layers later.

File: 3rdparty/mshadow/mshadow/base.h [hunk 10]
```diff
@@ -988,6 +1034,7 @@ struct minimum {
 };
 }  // namespace red
 
+#ifndef __NVCC__
```
Comment (ptrendx) [inline 3rdparty/mshadow/mshadow/base.h]: I don't like it - can we make a similar thing as in the case of fp16 for CPUs that did not support F16C instructions (i.e. code that runs but may be slower than the code for hardware natively supporting bfloat16)?

Comment (ptrendx) [reply 3rdparty/mshadow/mshadow/base.h]: You can implement `atomicAdd` (which seems to be the problem you are facing)with `atomicCAS` like this: https://github.com/apache/incubator-mxnet/blob/master/src/common/cuda_utils.h#L702-L721

**Comment by eric-haibin-lin** [comment]
@ElaineBao thanks for the explanation

File: 3rdparty/mshadow/mshadow/base.h (hunk 10 above)
Comment (ZhennanQin) [reply 3rdparty/mshadow/mshadow/base.h]: We don't have enough background / knowledge to enable Bfloat16 on GPU side. So probably we can't make the change you proposed. Alternately, any code refactoring on GPU side is welcome. you may change this in following PR.

File: python/mxnet/gluon/parameter.py (hunk 6 above)
Comment (rongzha1) [reply python/mxnet/gluon/parameter.py]: np.dtype(self.dtype) is different from np.dtype(self.dtype).type
https://docs.scipy.org/doc/numpy/reference/generated/numpy.dtype.type.html
https://docs.scipy.org/doc/numpy/reference/generated/numpy.dtype.html#numpy.dtype

File: .gitmodules [hunk 11]
```diff
@@ -6,7 +6,7 @@
 	url = https://github.com/dmlc/ps-lite
 [submodule "3rdparty/dlpack"]
 	path = 3rdparty/dlpack
-	url = https://github.com/dmlc/dlpack
+	url = https://github.com/dmlc/dlpack.git
```
Comment (xinyu-intel) [inline .gitmodules]: keep same.

File: src/operator/nn/fully_connected.cc [hunk 12]
```diff
@@ -237,7 +238,7 @@ static bool BackwardFCStorageType(const nnvm::NodeAttrs& attrs,
   bool dispatched = false;
   if (!dispatched && common::ContainsOnlyStorage(*in_attrs, mxnet::kDefaultStorage)) {
     dispatched = storage_type_assign(out_attrs, mxnet::kDefaultStorage,
-                                     dispatch_mode, DispatchMode::kFCompute);
+                                     dispatch_mode, DispatchMode::kFComputeEx);
```
Comment (xinyu-intel) [inline src/operator/nn/fully_connected.cc]: we may need to enable dnnl fc bwd in another PR since there is an known issue.

File: .gitmodules (hunk 11 above)
Comment (rongzha1) [reply .gitmodules]: OK

File: src/operator/nn/fully_connected.cc (hunk 12 above)
Comment (rongzha1) [reply src/operator/nn/fully_connected.cc]: Thanks for reminder

**Comment by pengzhao-intel** [comment]
@ptrendx thanks for your review.  Feel free to let me know if you have other concerns, we are going to merge PR recently.

**Comment by TaoLv** [comment]
Hi @leezu, @larroy, this PR can pass the builds for ARM but always hit the time out of the test. 
http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/mxnet-validation%2Fedge/detail/PR-17265/29/pipeline

We don't have any environment to reproduce. Could you please take a look or have any suggestion for further debugging?

File: 3rdparty/mshadow/mshadow/bfloat.h [hunk 13]
```diff
@@ -0,0 +1,167 @@
+/*!
```
Comment (TaoLv) [inline 3rdparty/mshadow/mshadow/bfloat.h]: @szha Do we need Apache license header for this new file?

Comment (szha) [reply 3rdparty/mshadow/mshadow/bfloat.h]: yes

**Comment by zhreshold** [comment]
@larroy Do you have idea how to display more logs for the edge tests? It consistently fail at this stage.

File: 3rdparty/mshadow/mshadow/bfloat.h (hunk 13 above)
Comment (rongzha1) [reply 3rdparty/mshadow/mshadow/bfloat.h]: done  thanks

**Comment by pengzhao-intel** [comment]
I am merging now. If there're any other comments, we can resolve by new PR.
